### PR TITLE
Windows permissions

### PR DESCRIPTION
**Pull request by sonologic**
Added code to auto-set core.filemode based on file-system probe (as is done in cgit), added a number of test-cases as well.

Also fixed a number of test-cases failing on windows.

### CODE REVIEW

File: dulwich/tests/compat/test_repository.py [hunk 1]
```diff
@@ -179,13 +181,21 @@ def test_git_worktree_list(self):
         worktrees = self._parse_worktree_list(output)
         self.assertEqual(len(worktrees), self._number_of_working_tree)
         self.assertEqual(worktrees[0][1], '(bare)')
-        self.assertEqual(worktrees[0][0], self._mainworktree_repo.path)
+        if sys.platform == 'win32':
+            # on windows, paths are case insensitive, and native git reports paths with '/' as seperator
```
Comment (jelmer) [inline dulwich/tests/compat/test_repository.py]: Please use os.path.normcase here - that should remove the need for Windows-specific code. normcase is a no-op on POSIX.

**Comment by sonologic** [comment]
I'll test the new commit (normcase) on windows tomorrow when @ work.

**Comment by sonologic** [comment]
I'll also fix the SwitftRepo test-case. No clue what that is but apparently those test-cases are skipped on my setups..

**Comment by sonologic** [comment]
Ok, test-cases pass on windows too after the changes.

On 2 January 2017 20:01:17 CET, "Jelmer Vernooĳ" <notifications@github.com> wrote:
>jelmer commented on this pull request.
>
>
>
>> @@ -179,13 +181,21 @@ def test_git_worktree_list(self):
>         worktrees = self._parse_worktree_list(output)
>         self.assertEqual(len(worktrees), self._number_of_working_tree)
>         self.assertEqual(worktrees[0][1], '(bare)')
>-        self.assertEqual(worktrees[0][0],
>self._mainworktree_repo.path)
>+        if sys.platform == 'win32':
>+            # on windows, paths are case insensitive, and native git
>reports paths with '/' as seperator
>
>Please use os.path.normcase here - that should remove the need for
>Windows-specific code. normcase is a no-op on POSIX.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/jelmer/dulwich/pull/472#pullrequestreview-14866641

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by jelmer** [comment]
On Tue, Jan 03, 2017 at 01:24:18AM -0800, sonologic wrote:
> Ok, test-cases pass on windows too after the changes.
Thanks, LGTM. I'll have a look at merging this this evening.

**Comment by jelmer** [comment]
Merged, thanks!

**Comment by koenfugro** [comment]
Thank you for accepting the changes, makes my life easier :)

However, I fear something has gone awry with the NEWS file: it now looks like the changes were released with 0.16.0, but they're not. I guess they'll be in 0.16.2 ?

**Comment by jelmer** [comment]
On Wed, Jan 04, 2017 at 02:31:10AM -0800, koenfugro wrote:
> Thank you for accepting the changes, makes my life easier :)
> 
> However, I fear something has gone awry with the NEWS file: it now looks like the changes were released with 0.16.0, but they're not. I guess they'll be in 0.16.2 ?
Ah, yep. Fix pushed.

**Comment by koenfugro** [comment]
Seems like your refactoring of line ~511 in index.py broke something. It still passes test-cases, but i'm now getting a big fat stack trace when using your version:

Traceback (most recent call last):
  File "tests\test_git_repository.py", line 63, in test_checkout
    repo.checkout('rel-2032031101')
  File "git_repository.py", line 120, in checkout
    r.reset_index(r['refs/tags/' + tag].tree)
  File "C:\Python27\lib\site-packages\dulwich-0.16.2-py2.7.egg\dulwich\repo.py", line 934, in reset_index
    validate_path_element=validate_path_element)
  File "C:\Python27\lib\site-packages\dulwich-0.16.2-py2.7.egg\dulwich\index.py", line 514, in build_index_from_tree
    index.write()
  File "C:\Python27\lib\site-packages\dulwich-0.16.2-py2.7.egg\dulwich\index.py", line 220, in write
    write_index_dict(f, self._byname)
  File "C:\Python27\lib\site-packages\dulwich-0.16.2-py2.7.egg\dulwich\index.py", line 175, in write_index_dict
    write_index(f, entries_list)
  File "C:\Python27\lib\site-packages\dulwich-0.16.2-py2.7.egg\dulwich\index.py", line 165, in write_index
    write_cache_entry(f, x)
  File "C:\Python27\lib\site-packages\dulwich-0.16.2-py2.7.egg\dulwich\index.py", line 125, in write_cache_entry
    write_cache_time(f, ctime)
  File "C:\Python27\lib\site-packages\dulwich-0.16.2-py2.7.egg\dulwich\index.py", line 93, in write_cache_time
    raise TypeError(t)
TypeError: 1484050597

Problem is that the time becomes a long, whereas write_cache_time expexts either an int, a float or a tuple.

Not sure why the dulwich test-cases are passing though, will check it out, but apparently you version:

    st = st.__class__((entry.mode, ) + st[1:])

is not equivalent to mine:

    st_tuple = (entry.mode, st.st_ino, st.st_dev, st.st_nlink, st.st_uid, st.st_gid, st.st_size, st.st_atime, st.st_mtime, st.st_ctime)
    st = st.__class__(st_tuple)

**Comment by koenfugro** [comment]
Right right i remember again.. If you access the stat_result object as a tuple, it will represent the time as a long, while if you access it as st.st_atime (or mtime, ctime) you get the actual float instead of the long representation..

So the formatting change broke things, because it wasn't just formatting that's changed. It's actually changing the representation so introducing a bug. This is windows-specific behaviour btw.

I'm open to suggestions on how to deal with this properly, but my idea now is to add a new patch that restores the old st_tuple construction, so that the type of the times is preserved, and add a line of comment that it is intentionally not written as tuple slicing...?

**Comment by koenfugro** [comment]
Btw, running the dulwich test_suite on windows with the 'formatting change' results in lots of failing test-cases with the exact same TypeError now.

**Comment by koenfugro** [comment]
Created new issue for this: https://github.com/jelmer/dulwich/issues/478 (because this one is closed and I can't re-open)

**Pull request closed without merging**